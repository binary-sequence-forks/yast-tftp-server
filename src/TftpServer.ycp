/**
 * File:	modules/TftpServer.ycp
 * Package:	Configuration of TftpServer
 * Summary:	Data for configuration of TftpServer, input and output functions.
 * Authors:	Martin Vidner <mvidner@suse.cz>
 *
 * $Id$
 *
 * Representation of the configuration of TftpServer.
 * Input and output routines.
 */

{

module "TftpServer";
textdomain "tftp-server";

import "Runlevel";
import "Report";
import "Summary";

/**
 * Required packages for operation
 */
global list required_packages = ["xinetd", "tftp"];

/**
 * Start tftpd via xinetd?
 */
global boolean start = false;

/**
 * Image directory, last argument of in.tftpd
 */
global string directory = "";

/**
 * Other arguments to in.tftpd, ie. not including -s or /dir
 */
global string other_args = "";


/**
 * Detect who is serving tftp:
 * Inetd may be running, it is the default. But it is ok unless it is
 * serving tftp. So we detect who is serving tftp and warn if it is
 * not xinetd or in.tftpd.
 * If nonempty, the user is notified and the module gives up.
 */
global string foreign_servers = "";

/**
 * Extract the directory and other arguments.
 * global to make testing easier
 * @param server_args server_args from xinetd.conf
 */
global define void ParseServerArgs (string server_args) ``{
    // extract the last argument and kick "-s".
    list(string) server_args_l = filter (string s,
					 splitstring (server_args, " \t"),
					 ``( s != ""));
    integer sz = size (server_args_l);
    integer i = 0;
    list(string) other_args_l = filter (string s, server_args_l, ``{
	i = i + 1;
	return s != "-s" && i != sz;
    });
    directory = server_args_l[sz - 1]:"";
    other_args = mergestring (other_args_l, " ");
}

/**
 * Read all tftp-server settings
 * @return true on success
 */
global define boolean Read () ``{
    // foreign_servers:
    // get command names via lsof, filter out xinetd and in.tftpd
    map out = SCR::Execute (.target.bash_output, "/usr/bin/lsof -i :tftp -Fc");
    list(string) lines = splitstring (out["stdout"]:"", "\n");
    // the command is field 'c'
    lines = filter (string l, lines, ``( substring (l, 0, 1) == "c" ));
    // strip the c
    lines = maplist (string l, lines, ``( substring (l, 1, size (l) - 1) ));
    // filter out our servers
    lines = filter (string l, lines, ``( l != "xinetd" && l != "in.tftpd" ));
    foreign_servers = mergestring (lines, ", ");

    boolean xinetd_start = Runlevel::ServiceEnabled ("xinetd");

    // is the config file there at all?
    list(string) sections = SCR::Dir (.etc.xinetd_d.tftp.section);
    string disable = SCR::Read (.etc.xinetd_d.tftp.value.tftp.disable);
    start = xinetd_start && sections != [] && disable != "yes";

    string server_args = SCR::Read (.etc.xinetd_d.tftp.value.tftp.server_args);
    if (server_args == nil)
    {
	// default
//	server_args = "-s /tftpboot";
	server_args = "";
    }

    ParseServerArgs (server_args);

    return true;
}

/**
 * Return error string to be used in WriteOnly (for autoinst)
 * or before the edit dialog.
 * @return error string
 */
global define string ForeignServersError () ``{
    // error popup
    // %1 is a command name (or a comma (, ) separated list of them)
    return sformat (_("This module can only use xinetd to set up TFTP.
However, another program is serving TFTP: %1.
Exiting.
"), foreign_servers);
}

/**
 * Write all tftp-server settings
 * without actually (re)starting the service
 * @return true on success
 */
global define boolean WriteOnly () ``{
    y2milestone ("Writing");

    // give up if tftp is served by someone else
    if (foreign_servers != "")
    {
	Report::Error (ForeignServersError ());
	return false;
    }

    // write the config file
    //
    //  create it if it does not exist
    //  could be a normal situation at initial setup
    //  or a broken setup, ok if we fix it when writing
    //  but that means messing up with other parameters
    //  lets touch just the basics
    //  the first "item" is the brace following the section start
    SCR::Write (.etc.xinetd_d.tftp.value.tftp."{", "");
    SCR::Write (.etc.xinetd_d.tftp.value_type.tftp."{", 1);
    SCR::Write (.etc.xinetd_d.tftp.value.tftp.disable, (start? "no": "yes"));
    if (start)
    {
	SCR::Write (.etc.xinetd_d.tftp.value.tftp.socket_type, "dgram");
	SCR::Write (.etc.xinetd_d.tftp.value.tftp.protocol, "udp");
	SCR::Write (.etc.xinetd_d.tftp.value.tftp.wait, "yes");
	SCR::Write (.etc.xinetd_d.tftp.value.tftp.user, "root");
	SCR::Write (.etc.xinetd_d.tftp.value.tftp.server, "/usr/sbin/in.tftpd");
	string server_args = sformat ("%1 -s %2", other_args, directory);
	SCR::Write (.etc.xinetd_d.tftp.value.tftp.server_args, server_args);
    }

    // flush
    SCR::Write (.etc.xinetd_d.tftp, nil);

    // image dir: if does not exist, create with root:root rwxr-xr-x
    SCR::Execute (.target.mkdir, directory);

    // firewall??

    // enable and (re)start xinetd
    if (start)
    {
	Runlevel::ServiceAdjust ("xinetd", "enable");
    }
    else
    {
	// xinetd may be needed for other services so we never turn it
	// off.
    }

    return true;
}

/**
 * Write all tftp-server settings
 * @return true on success
 */
global define boolean Write () ``{
    // write the config file
    // image dir: if does not exist, create with root:root rwxr-xr-x
    // firewall??
    // enable and (re)start xinetd

    if (!WriteOnly ())
    {
	return false;
    }

    // enable and (re)start xinetd

    // in.tftpd will linger around for 15 minutes waiting for a new connection
    // so we must kill it otherwise it will be using the old parameters
    SCR::Execute (.target.bash, "/usr/bin/killall in.tftpd");

    if (start)
    {
	Runlevel::RunInitScript ("xinetd", "restart");
    }
    else
    {
	// xinetd may be needed for other services so we never turn it
	// off. It will exit anyway if no services are configured.
	// If it is running, restart it.
	Runlevel::RunInitScript ("xinetd", "try-restart");
    }

    return true;
}

/**
 * Set module data, without validity checking
 * @param settings may be empty for reset
 * @return void
 */
global define void Set (map settings) ``{
    start = settings["start_tftpd"]:false;
    directory = settings["tftp_directory"]:"";
    other_args = "";
}


/**
 * Get all tftp-server settings from the first parameter
 * (For use by autoinstallation.)
 * @param settings The YCP structure to be imported.
 * @return boolean True on success
 */
global define boolean Import (map settings) ``{
    if (settings != $[])
    {
	string s = "start_tftpd";
	string d = "tftp_directory";
	if (!haskey (settings, s))
	{
	    y2error ("Missing at Import: '%1'.", s);
	    return false;
	}
	if (settings[s]:false && !haskey (settings, d))
	{
	    y2error ("Missing at Import: '%1'.", d);
	    return false;
	}
    }
    Set (settings);
    return true;
}

/**
 * Dump the tftp-server settings to a single map
 * (For use by autoinstallation.)
 * @return map Dumped settings (later acceptable by Import ())
 */
global define map Export () ``{
    map settings = $[
	"start_tftpd": start,
	];
    if (start)
    {
	settings["tftp_directory"]= directory;
    }
    return settings;
}

/**
 * @return Html formatted configuration summary
 */
global define string Summary () ``{
    string summary = "";
    string nc = Summary::NotConfigured ();

    // summary header
    summary = Summary::AddHeader(summary, _("TFTP Server Enabled"));
    // summary item: an option is turned on
    summary = Summary::AddLine(summary, (start) ? _("Yes") : nc);

    // summary header
    summary = Summary::AddHeader(summary, _("Boot Image Directory"));
    summary = Summary::AddLine(summary, (directory != "") ? directory : nc);

    return summary;
}

/**
 * Return needed packages and packages to be removed 
 * during autoinstallation.
 * @return map of lists.
 *
**/

global  define map AutoPackages() ``{
   list install_pkgs = required_packages;
   list remove_pkgs = [];
   return ($["install": install_pkgs, "remove": remove_pkgs]);
}



/* EOF */
}
