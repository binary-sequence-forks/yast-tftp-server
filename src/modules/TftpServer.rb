# encoding: utf-8

# File:	modules/TftpServer.ycp
# Package:	Configuration of TftpServer
# Summary:	Data for configuration of TftpServer, input and output functions.
# Authors:	Martin Vidner <mvidner@suse.cz>
#
# $Id$
#
# Representation of the configuration of TftpServer.
# Input and output routines.
require "yast"

module Yast
  class TftpServerClass < Module
    def main
      textdomain "tftp-server"

      Yast.import "Progress"
      Yast.import "Report"
      Yast.import "Service"
      Yast.import "Summary"
      Yast.import "SuSEFirewall"

      # Any settings modified?
      # As we have only a single dialog which handles it by itself,
      # it is used only by autoinst cloning.
      @modified = false

      # Required packages for operation
      @required_packages = ["xinetd", "tftp"]

      # Start tftpd via xinetd?
      @start = false

      # Image directory, last argument of in.tftpd
      @directory = ""

      # Other arguments to in.tftpd, ie. not including -s or /dir
      @other_args = ""


      # Detect who is serving tftp:
      # Inetd may be running, it is the default. But it is ok unless it is
      # serving tftp. So we detect who is serving tftp and warn if it is
      # not xinetd or in.tftpd.
      # If nonempty, the user is notified and the module gives up.
      @foreign_servers = ""
    end

    # Returns true if the settings were modified
    # @return settings were modified
    def GetModified
      @modified
    end

    # Function sets an internal variable indicating that any
    # settings were modified to "true".
    # Used for autoinst cloning.
    def SetModified
      @modified = true

      nil
    end

    # Extract the directory and other arguments.
    # global to make testing easier
    # @param [String] server_args server_args from xinetd.conf
    def ParseServerArgs(server_args)
      # extract the last argument and kick "-s".
      server_args_l = Builtins.filter(Builtins.splitstring(server_args, " \t")) do |s|
        s != ""
      end
      sz = Builtins.size(server_args_l)
      i = 0
      other_args_l = Builtins.filter(server_args_l) do |s|
        i = Ops.add(i, 1)
        s != "-s" && i != sz
      end
      @directory = Ops.get(server_args_l, Ops.subtract(sz, 1), "")
      @other_args = Builtins.mergestring(other_args_l, " ")

      nil
    end

    # Read all tftp-server settings
    # @return true on success
    def Read
      # foreign_servers:
      # get command names via lsof, filter out xinetd and in.tftpd
      out = Convert.to_map(
        SCR.Execute(path(".target.bash_output"), "/usr/bin/lsof -i :tftp -Fc")
      )
      lines = Builtins.splitstring(Ops.get_string(out, "stdout", ""), "\n")
      # the command is field 'c'
      lines = Builtins.filter(lines) { |l| Builtins.substring(l, 0, 1) == "c" }
      # strip the c
      lines = Builtins.maplist(lines) do |l|
        Builtins.substring(l, 1, Ops.subtract(Builtins.size(l), 1))
      end
      # filter out our servers
      lines = Builtins.filter(lines) { |l| l != "xinetd" && l != "in.tftpd" }
      @foreign_servers = Builtins.mergestring(lines, ", ")

      xinetd_start = Service.Enabled("xinetd")

      # is the config file there at all?
      sections = SCR.Dir(path(".etc.xinetd_d.tftp.section"))
      disable = Convert.to_string(
        SCR.Read(path(".etc.xinetd_d.tftp.value.tftp.disable"))
      )
      @start = xinetd_start && sections != [] && disable != "yes"

      server_args = Convert.to_string(
        SCR.Read(path(".etc.xinetd_d.tftp.value.tftp.server_args"))
      )
      if server_args == nil
        # default
        #	server_args = "-s /tftpboot";
        server_args = ""
      end

      ParseServerArgs(server_args)

      # TODO only when we have our own Progress
      #boolean progress_orig = Progress::set (false);
      SuSEFirewall.Read
      #Progress::set (progress_orig);

      true
    end

    # Return error string to be used in WriteOnly (for autoinst)
    # or before the edit dialog.
    # @return error string
    def ForeignServersError
      # error popup
      # %1 is a command name (or a comma (, ) separated list of them)
      Builtins.sformat(
        _(
          "This module can only use xinetd to set up TFTP.\n" +
            "However, another program is serving TFTP: %1.\n" +
            "Exiting.\n"
        ),
        @foreign_servers
      )
    end

    # Write all tftp-server settings
    # without actually (re)starting the service
    # @return true on success
    def WriteOnly
      Builtins.y2milestone("Writing")

      # give up if tftp is served by someone else
      if @foreign_servers != ""
        Report.Error(ForeignServersError())
        return false
      end

      # write the config file
      #
      #  create it if it does not exist
      #  could be a normal situation at initial setup
      #  or a broken setup, ok if we fix it when writing
      #  but that means messing up with other parameters
      #  lets touch just the basics
      #  the first "item" is the brace following the section start
      SCR.Write(path(".etc.xinetd_d.tftp.value.tftp.\"{\""), "")
      SCR.Write(path(".etc.xinetd_d.tftp.value_type.tftp.\"{\""), 1)
      SCR.Write(
        path(".etc.xinetd_d.tftp.value.tftp.disable"),
        @start ? "no" : "yes"
      )
      if @start
        SCR.Write(path(".etc.xinetd_d.tftp.value.tftp.socket_type"), "dgram")
        SCR.Write(path(".etc.xinetd_d.tftp.value.tftp.protocol"), "udp")
        SCR.Write(path(".etc.xinetd_d.tftp.value.tftp.wait"), "yes")
        SCR.Write(path(".etc.xinetd_d.tftp.value.tftp.user"), "root")
        SCR.Write(
          path(".etc.xinetd_d.tftp.value.tftp.server"),
          "/usr/sbin/in.tftpd"
        )
        server_args = Builtins.sformat("%1 -s %2", @other_args, @directory)
        SCR.Write(
          path(".etc.xinetd_d.tftp.value.tftp.server_args"),
          server_args
        )
      end

      # flush
      SCR.Write(path(".etc.xinetd_d.tftp"), nil)

      # image dir: if does not exist, create with root:root rwxr-xr-x
      SCR.Execute(path(".target.mkdir"), @directory)

      # firewall??

      # enable and (re)start xinetd
      Service.Enable("xinetd") if @start

      # TODO only when we have our own Progress
      #boolean progress_orig = Progress::set (false);
      SuSEFirewall.WriteOnly
      #Progress::set (progress_orig);

      true
    end

    # Write all tftp-server settings
    # @return true on success
    def Write
      # write the config file
      # image dir: if does not exist, create with root:root rwxr-xr-x
      # firewall??
      # enable and (re)start xinetd

      return false if !WriteOnly()

      # enable and (re)start xinetd

      # in.tftpd will linger around for 15 minutes waiting for a new connection
      # so we must kill it otherwise it will be using the old parameters
      SCR.Execute(path(".target.bash"), "/usr/bin/killall in.tftpd")

      if @start
        Service.Restart("xinetd")
      else
        # xinetd may be needed for other services so we never turn it
        # off. It will exit anyway if no services are configured.
        # If it is running, restart it.
        Service.RunInitScript("xinetd", "try-restart")
      end

      # TODO only when we have our own Progress
      #boolean progress_orig = Progress::set (false);
      SuSEFirewall.ActivateConfiguration
      #Progress::set (progress_orig);

      true
    end

    # Set module data, without validity checking
    # @param [Hash] settings may be empty for reset
    # @return [void]
    def Set(settings)
      settings = deep_copy(settings)
      @start = Ops.get_boolean(settings, "start_tftpd", false)
      @directory = Ops.get_string(settings, "tftp_directory", "")
      @other_args = ""

      nil
    end


    # Get all tftp-server settings from the first parameter
    # (For use by autoinstallation.)
    # @param [Hash] settings The YCP structure to be imported.
    # @return [Boolean] True on success
    def Import(settings)
      settings = deep_copy(settings)
      if settings != {}
        s = "start_tftpd"
        d = "tftp_directory"
        if !Builtins.haskey(settings, s)
          Builtins.y2error("Missing at Import: '%1'.", s)
          return false
        end
        if Ops.get_boolean(settings, s, false) && !Builtins.haskey(settings, d)
          Builtins.y2error("Missing at Import: '%1'.", d)
          return false
        end
      end
      Set(settings)
      true
    end

    # Dump the tftp-server settings to a single map
    # (For use by autoinstallation.)
    # @return [Hash] Dumped settings (later acceptable by Import ())
    def Export
      settings = { "start_tftpd" => @start }
      Ops.set(settings, "tftp_directory", @directory) if @start
      deep_copy(settings)
    end

    # @return Html formatted configuration summary
    def Summary
      summary = ""
      nc = Summary.NotConfigured

      # summary header
      summary = Summary.AddHeader(summary, _("TFTP Server Enabled:"))
      # summary item: an option is turned on
      summary = Summary.AddLine(summary, @start ? _("Yes") : nc)

      # summary header
      summary = Summary.AddHeader(summary, _("Boot Image Directory:"))
      summary = Summary.AddLine(summary, @directory != "" ? @directory : nc)

      summary
    end

    # Return needed packages and packages to be removed
    # during autoinstallation.
    # @return [Hash] of lists.
    #
    #

    def AutoPackages
      install_pkgs = deep_copy(@required_packages)
      remove_pkgs = []
      { "install" => install_pkgs, "remove" => remove_pkgs }
    end

    publish :function => :GetModified, :type => "boolean ()"
    publish :function => :SetModified, :type => "void ()"
    publish :variable => :required_packages, :type => "list <string>"
    publish :variable => :start, :type => "boolean"
    publish :variable => :directory, :type => "string"
    publish :variable => :other_args, :type => "string"
    publish :variable => :foreign_servers, :type => "string"
    publish :function => :ParseServerArgs, :type => "void (string)"
    publish :function => :Read, :type => "boolean ()"
    publish :function => :ForeignServersError, :type => "string ()"
    publish :function => :WriteOnly, :type => "boolean ()"
    publish :function => :Write, :type => "boolean ()"
    publish :function => :Set, :type => "void (map)"
    publish :function => :Import, :type => "boolean (map)"
    publish :function => :Export, :type => "map ()"
    publish :function => :Summary, :type => "string ()"
    publish :function => :AutoPackages, :type => "map ()"
  end

  TftpServer = TftpServerClass.new
  TftpServer.main
end
