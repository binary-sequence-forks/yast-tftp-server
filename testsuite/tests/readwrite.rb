# encoding: utf-8

# Module:  TFTP server configuration
# Summary: Testsuite
# Authors: Martin Vidner <mvidner@suse.cz>
#
# $Id$
module Yast
  class ReadwriteClient < Client
    def main
      # testedfiles: TftpServer.ycp Service.ycp Testsuite.ycp

      Yast.include self, "testsuite.rb"

      @READ = {
        "etc"    => {
          "xinetd_d" => {
            "tftp" => {
              "section" => { "tftp" => {} },
              "value"   => {
                "tftp" => {
                  "disable"     => nil,
                  "server_args" => "-s -v /srv/boot"
                }
              }
            }
          }
        },
        "target" => { "size" => 0 }
      }

      @WRITE = {}

      @EXECUTE = {
        "target" =>
          # 	    "remove": true, // /etc/yp.conf.sv
          {
            # ok if used both for `lsof` and `rcfoo start`
            "bash_output" => {
              "exit"   => 0,
              "stderr" => "",
              "stdout" => "p3316\ncxinetd\n"
            }
          }
      }

      TESTSUITE_INIT([@READ, @WRITE, @EXECUTE], nil)

      Yast.import "TftpServer"

      TEST(lambda { TftpServer.Read }, [@READ, @WRITE, @EXECUTE], nil)
      TEST(lambda { TftpServer.Write }, [@READ, @WRITE, @EXECUTE], nil) 

      #READ["etc","xinetd_d","tftp","value","tftp","server_args"] = ...

      nil
    end
  end
end

Yast::ReadwriteClient.new.main
