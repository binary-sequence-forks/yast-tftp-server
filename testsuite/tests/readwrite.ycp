/**
 * Module:  TFTP server configuration
 * Summary: Testsuite
 * Authors: Martin Vidner <mvidner@suse.cz>
 *
 * $Id$
 */
{
    // testedfiles: TftpServer.ycp Service.ycp Testsuite.ycp

    include "testsuite.ycp";

    map READ = $[
	// Runlevel:
	"init": $[
	    "scripts": $[
		"exists": true,
		"runlevel": $[
		    "inetd": $[
			"start": [],
			"stop":	 [],
			],
		    "xinetd": $[
			"start": [ "3", "5"],
			"stop":	 [ "3", "5"],
			],
		    ],
		// their contents is not important for ServiceAdjust
		"comment": $[
		    "inetd": $[],
		    "xinetd": $[],
		    ],
		],
	    ],
	"etc": $[
	    "xinetd_d": $[
		"tftp": $[
		    "section": $[
			"tftp": $[],
			],
		    "value": $[
			"tftp": $[
			    "disable": nil,
			    "server_args": "-s -v /srv/boot"
			    ],
			],
		    ],
		],
	    ],
	    "target" : $[
		"size" : 0,
	    ],
	];

    map WRITE = $[
	];

    map EXECUTE = $[
	"target": $[
	    // ok if used both for `lsof` and `rcfoo start`
	    "bash_output": $[
		"exit":0,
		"stderr":"",
		"stdout":"p3316\ncxinetd\n"
		],
/*
	    "remove": true, // /etc/yp.conf.sv
*/
	    ],
	];

    TESTSUITE_INIT ([READ, WRITE, EXECUTE], nil);

    import "TftpServer";

    TEST (``(TftpServer::Read ()), [READ, WRITE, EXECUTE], nil);
    TEST (``(TftpServer::Write ()), [READ, WRITE, EXECUTE], nil);

    //READ["etc","xinetd_d","tftp","value","tftp","server_args"] = ...
}
