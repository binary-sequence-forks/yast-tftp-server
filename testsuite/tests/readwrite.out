Execute	.target.bash_output "/usr/bin/lsof -i :tftp -Fc" $["exit":0, "stderr":"", "stdout":"p3316\ncxinetd\n"]
Execute	.target.bash "/sbin/chkconfig -l xinetd | /usr/bin/grep -q '[0-6]:on'" 0
Dir	.etc.xinetd_d.tftp.section: ["tftp"]
Read	.etc.xinetd_d.tftp.value.tftp.disable nil
Read	.etc.xinetd_d.tftp.value.tftp.server_args "-s -v /srv/boot"
Execute	.target.bash "/sbin/chkconfig -l SuSEfirewall2_init | /usr/bin/grep -q '[0-6]:on'" 0
Execute	.target.bash "/sbin/chkconfig -l SuSEfirewall2_setup | /usr/bin/grep -q '[0-6]:on'" 0
Execute	.target.bash "/sbin/service SuSEfirewall2_init status" $["TERM":"raw"] 0
Return	true
Write	.etc.xinetd_d.tftp.value.tftp."{" "" true
Write	.etc.xinetd_d.tftp.value_type.tftp."{" 1 true
Write	.etc.xinetd_d.tftp.value.tftp.disable "no" true
Write	.etc.xinetd_d.tftp.value.tftp.socket_type "dgram" true
Write	.etc.xinetd_d.tftp.value.tftp.protocol "udp" true
Write	.etc.xinetd_d.tftp.value.tftp.wait "yes" true
Write	.etc.xinetd_d.tftp.value.tftp.user "root" true
Write	.etc.xinetd_d.tftp.value.tftp.server "/usr/sbin/in.tftpd" true
Write	.etc.xinetd_d.tftp.value.tftp.server_args "-v -s /srv/boot" true
Write	.etc.xinetd_d.tftp nil true
Execute	.target.mkdir "/srv/boot" 0
Read	.init.scripts.runlevel "xinetd" $["inetd":$["start":[], "stop":[]], "network":$["start":[], "stop":[]], "networkmanager":$["start":["3", "5"], "stop":["3", "5"]], "xinetd":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.init.scripts.comment "xinetd" $["inetd":$[], "network":$[], "networkmanager":$[], "xinetd":$[]]
Execute	.target.bash "/usr/bin/killall in.tftpd" 0
Execute	.target.bash_output "/sbin/service xinetd restart" $["TERM":"raw"] $["exit":0, "stderr":"", "stdout":"p3316\ncxinetd\n"]
Execute	.target.bash "/sbin/service SuSEfirewall2_init status" $["TERM":"raw"] 0
Return	true
